lib/SIL/Utils/DynamicCasts.cpp

```diff
@@ -208,6 +208,41 @@ static CanType getHashableExistentialType(ModuleDecl *M) {
   return hashable->getDeclaredInterfaceType()->getCanonicalType();
 }
 
+bool SILDynamicCastInst::isRCIdentityPreserving() const {
+  // Casts which cast from a trivial type, like a metatype, to something which
+  // is retainable, like an AnyObject, are not RC identity preserving.
+  // On some platforms such casts dynamically allocate a ref-counted box for the
+  // metatype. Naturally that is the place where a new rc-identity begins.
+  // Therefore such a cast is introducing a new rc identical object.
+  //
+  // If RCIdentityAnalysis would look through such a cast, ARC optimizations
+  // would get confused and might eliminate a retain of such an object
+  // completely.
+  if (getSourceLoweredType().isTrivial(*getFunction()))
+    return false;
+
+  CanType source = getSourceFormalType();
+  CanType target = getTargetFormalType();
+
+  // Be conservative with generic parameters.
+  if (source->is<ArchetypeType>() || target->is<ArchetypeType>())
```

Comment:
The `ArchetypeType` check is confusing to me, and "be conservative" doesn't help. In this context, based on all the comments so far, the check implies that an archetype can be substituted with a Metatype. I don't think that true, is it? If so, what would that look like? Instead, the comment should indicate specifically what kind of cast we need to protect against here.

Or is it simply that one of the Archetypes may be a bridged type and one side is non-bridged? In that case, why don't we have the same problem with casting concrete types? (can a concrete bridgeable struct be cast to a bridged class?)

Why don't we treat Archetype's similar to existentials and allow the RC-preserving cast as long as `requiresClass()` is true for both sides?

Reply:
I wonder if we could maybe have a different instruction or something like that. The problem here is as I understand it around bridging. Rather than using normal casts for bridging, lets use special casts. Then we have a nice split.

Reply:
I think whether a cast is bridging depends entirely on the types and may depend on substituted types. I mainly want to know the rules for which type classifications may result in bridging (metatype to ObjC class, struct to ObjC object, ...), whether it's legal to substitute those types with archetypes. Maybe @slavapestov, @tbkka, or @Catfish-Man can help classify these dynamic casts.

For example, in a struct-to-ObjC-object cast, can we assume that the struct is always a trivial type? If so, we just need to comment the 'isTrivial' check to make it clear that it covers bridged structs too, not just metatypes. If not, it seems that we're still missing a case here.

Reply:
Remind me what Swift considers trivial? Is String trivial? I’d assume not...

Reply:
@Catfish-Man "trivial" basically means: no ref-counting involved, i.e. values can be copied with a simple memcpy. So yes, String is not trivial.

Reply:
@atrick I pushed a new version, which should make things clearer about archetypes. The idea is that an archetype can be a class or existential - we don't know.

This check is there to be absolutely on the safe side. Currently we even cannot hit this case, because the only cast instruction which can have an archetype is checked_cast_addr_br - which has the MayRelease attribute. And for RCidentity we don't deal with addr-casts.
I added the check to be future prove, once we switch to opaque values.

For the remaining logic (existential <-> class casts) I consulted with @tbkka (who rewrote big parts of the runtime cast machinery). This should catch all the cases where we can end up with not having RC identity.

Reply:
This change came out of some tests I'm developing to stress the cast machinery.  My tests uncovered problems with casts where one side is `AnyObject` and the other is a pointer type (metatype, class type, existential, etc).  These cases were failing because the code was implicitly assuming that such casts simply copied a pointer, but my tests demonstrate cases where the `AnyObject` references a container structure that in turn holds the pointer in question.  In these cases, retain/release of the `AnyObject` is not equivalent to retain/release on the contained pointer.

In particular, the previous logic could do the wrong thing in cases like the following:
```
   let source : AnyObject = ....
   // retain source for later use
  let result = source as? ClassType
  // Re-use source here
```
If `source` above is just a pointer to a class instance, then it is valid to sink the retain past the cast:
```
  let source : AnyObject = ClassType(). // <<== Simple case
  let result = source as? ClassType
  // retain source (which is the same pointer as result). <== THIS IS OK
  // Re-use source
```
But this transformation is not valid if `source` holds a pointer to a runtime container structure that in turn holds the class reference.  In that case, the transformation above releases the container prematurely:
```
  let source : AnyObject = ... something gotten from somewhere ...
  // MUST retain `source` BEFORE the cast, since source may not be a simple class reference
  let result = source as? ClassType
  // Re-use source
```

This is primarily a concern when there are complex interactions with Obj-C code.  The bridging support wraps Swift types into a variety of heap-allocated containers.  The compiler cannot productively reason about these.  They are being exchanged with Obj-C code that is not visible to the compiler, so their structure can really only be understood by dynamic inspection in the runtime casting code.

Reply:
@eeckstein @tbkka The code is more clear now. Thanks for your patience. It's always helpful for me to see more explanation like this in comments.

Just one question: we seem to be assuming that AnyObject -> Metatype casts will never happen because we don't check for trivial targets. If so, can we add that assertion at least?

Reply:
Oh?   AnyObject -> Metatype casts can definitely happen.  (Briefly: _anything_ can be stored in an `AnyObject` via casting and can later be pulled back out with another cast.)